File: rust/crates/spore-core/src/tools/params.rs
//! Per-tool serde input structs and parsing helper.

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

use crate::model::ToolCall;
use crate::tools::error::ToolExecutionError;

/// Parse `call.input` into a typed parameter struct. Any deserialization
/// failure is mapped to [`ToolExecutionError::InvalidParameters`].
pub fn parse_params<T: serde::de::DeserializeOwned>(
    call: &ToolCall,
) -> Result<T, ToolExecutionError> {
    serde_json::from_value::<T>(call.input.clone()).map_err(|e| {
        ToolExecutionError::InvalidParameters {
            reason: e.to_string(),
        }
    })
}

// ---------- Filesystem ----------

/// Parameters for [`crate::tools::fs::ReadFileTool`]. With all three optional
/// params at their defaults (`offset = 1`, `length = 0`, `line_numbers =
/// false`) the read is **byte-identical** to reading the whole file — no
/// header, no line numbers (#132). Any non-default param prepends a one-line
/// header `[lines {start}–{end} of {total}]` (U+2013 en-dash).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReadFileParams {
    pub path: String,
    /// 1-indexed start line. Default `1` (read from the beginning).
    #[serde(default = "default_read_offset")]
    pub offset: u64,
    /// Max lines to return. Default `0` = no limit (read to EOF). A `length`
    /// that runs past EOF silently returns through the last line.
    #[serde(default)]
    pub length: u64,
    /// When `true`, prefix each returned line with its 1-indexed number,
    /// right-padded to the digit-width of the file's total line count.
    #[serde(default)]
    pub line_numbers: bool,
}
fn default_read_offset() -> u64 {
    1
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WriteFileParams {
    pub path: String,
    pub content: String,
    #[serde(default)]
    pub append: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListDirParams {
    pub path: String,
    #[serde(default)]
    pub recursive: bool,
    /// When `false` (default), a recursive walk honors `.gitignore`/`.ignore`
    /// and skips VCS dirs (`.git/`) — keeping the listing focused on source and
    /// out of the way of build artifacts (`target/`, `node_modules/`). Set
    /// `true` to walk everything, ignored files included.
    #[serde(default)]
    pub include_ignored: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeleteFileParams {
    pub path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MoveFileParams {
    pub src: String,
    pub dst: String,
}

// ---------- Exec ----------

/// Parameters for the shell-free [`crate::tools::exec::ExecTool`]: a program
/// name plus verbatim argument vector. No shell is involved.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExecParams {
    pub command: String,
    #[serde(default)]
    pub args: Vec<String>,
    /// Timeout in whole seconds.
    #[serde(default)]
    pub timeout: Option<u64>,
}

/// Parameters for the real [`crate::tools::exec::BashCommandTool`]: a single
/// shell `script` run via `/bin/sh -c`, with an optional working directory.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ShellCommandParams {
    pub script: String,
    #[serde(default)]
    pub working_dir: Option<String>,
    /// Timeout in whole seconds.
    #[serde(default)]
    pub timeout: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RunTestsParams {
    pub command: String,
    pub working_dir: String,
    #[serde(default)]
    pub timeout: Option<u64>,
}

// ---------- Search ----------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GrepFilesParams {
    pub pattern: String,
    pub path: String,
    #[serde(default)]
    pub recursive: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FindFilesParams {
    pub glob: String,
    pub path: String,
}

// ---------- Git ----------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GitLogParams {
    #[serde(default = "default_log_n")]
    pub n: u32,
    #[serde(default = "default_log_format")]
    pub format: String,
}
fn default_log_n() -> u32 {
    20
}
fn default_log_format() -> String {
    "oneline".into()
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GitDiffParams {
    #[serde(default)]
    pub from: Option<String>,
    #[serde(default)]
    pub to: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GitCommitParams {
    pub message: String,
    #[serde(default)]
    pub files: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct GitStatusParams {}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GitResetParams {
    pub target: String,
    pub mode: super::git::GitResetMode,
}

// ---------- HTTP ----------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HttpGetParams {
    pub url: String,
    #[serde(default)]
    pub headers: Option<Map<String, Value>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HttpPostParams {
    pub url: String,
    pub body: Value,
    #[serde(default)]
    pub headers: Option<Map<String, Value>>,
}

// ---------- EditFile (#81, new) ----------

/// Parameters for [`crate::tools::edit::EditFileTool`]: replace the FIRST and
/// ONLY occurrence of `old_string` with `new_string` in the file at `path`.
/// The match must be unique — an absent or non-unique `old_string` is a
/// recoverable error.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EditFileParams {
    pub path: String,
    pub old_string: String,
    pub new_string: String,
}

// ---------- Grep (#81, new — output modes) ----------

/// Output modes for [`crate::tools::search::GrepTool`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum GrepOutputMode {
    /// Each matching line as `path:line:text` (default).
    #[default]
    Content,
    /// Distinct file paths that contain at least one match.
    FilesWithMatches,
    /// `path:count` for each file with matches.
    Count,
}

/// Parameters for the net-new [`crate::tools::search::GrepTool`]. Distinct from
/// [`GrepFilesParams`]: adds `output_mode` and `context_lines`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GrepParams {
    pub pattern: String,
    pub path: String,
    #[serde(default)]
    pub recursive: bool,
    #[serde(default)]
    pub output_mode: GrepOutputMode,
    /// Lines of context to show before and after each match (default: 0).
    /// When 0 the output is byte-identical to the pre-context behaviour.
    /// When > 0 the output uses standard `grep -C N` format:
    /// match lines use `:` separator, context lines use `-`, and non-adjacent
    /// groups are separated by a `--` line.
    #[serde(default)]
    pub context_lines: u32,
}

// ---------- SendMessage (#81, new) ----------

/// Parameters for [`crate::tools::message::SendMessageTool`].
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SendMessageParams {
    pub content: String,
}

// ---------- TodoWrite (#81, new) ----------

/// A single todo item managed by [`crate::tools::todo::TodoWriteTool`].
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct TodoItem {
    pub content: String,
    pub status: TodoStatus,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TodoStatus {
    Pending,
    InProgress,
    Completed,
}

/// Parameters for [`crate::tools::todo::TodoWriteTool`]: the agent supplies the
/// FULL desired todo list, which replaces the persisted list wholesale.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TodoWriteParams {
    pub todos: Vec<TodoItem>,
}

// ---------- WebFetch / WebSearch (#81, new) ----------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WebFetchParams {
    pub url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WebSearchParams {
    pub query: String,
}

// ---------- Tier 3: plan / clarify / abort (#81, new) ----------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EnterPlanModeParams {
    #[serde(default)]
    pub context: String,
}

/// Parameters for [`crate::tools::plan::ExitPlanModeTool`]. The agent supplies
/// the plan as a structured object that deserializes DIRECTLY into the existing
/// [`PlanArtifact`](crate::hooks::PlanArtifact) (issue #81, Q4a — no stub).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExitPlanModeParams {
    pub plan: crate::hooks::PlanArtifact,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AskUserQuestionParams {
    pub question: String,
    #[serde(default)]
    pub options: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AbortParams {
    pub reason: String,
}

// ---------- Subagent ----------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SubagentParams {
    pub instruction: String,
}

// ---------- TaskList (#71) ----------

/// Parameters for the [`crate::tools::tasklist::TaskListTool`], internally
/// tagged on `action`. Each variant carries exactly the fields that action
/// consumes.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "action", rename_all = "snake_case")]
pub enum TaskListParams {
    AddTask {
        description: String,
        /// Ids of tasks that must be `Completed` before this one runs (#118).
        /// Optional; defaults to empty. Validated by [`crate::tasklist::TaskList::add`].
        #[serde(default)]
        blockers: Vec<u32>,
    },
    UpdateTask {
        id: u32,
        #[serde(default)]
        status: Option<crate::tasklist::TaskStatus>,
        #[serde(default)]
        description: Option<String>,
    },
    CompleteTask {
        id: u32,
    },
    ListTasks {},
}

// ---------- Memory (#82) ----------

/// Parameters for the [`crate::tools::memory::MemoryTool`], internally tagged on
/// `operation`. `scope` is an explicit field on BOTH variants. `write` carries
/// the entry payload; `read` carries the recency `limit` and a `merged` flag.
///
/// `StorageScope::Local` is accepted by serde (so a bad-scope call reaches the
/// tool body) but rejected at runtime with a recoverable error — the advertised
/// schema enum omits `local`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "operation", rename_all = "snake_case")]
pub enum MemoryToolParams {
    Write {
        scope: crate::storage::StorageScope,
        role: String,
        content: String,
        /// Free-form metadata stored on the entry; defaults to `{}`.
        #[serde(default = "memory_empty_metadata")]
        metadata: Value,
    },
    Read {
        scope: crate::storage::StorageScope,
        /// When `true`, return the cross-scope merged view (User ∪ Project,
        /// newest-first, no dedup) instead of just `scope`. Defaults to `false`.
        #[serde(default)]
        merged: bool,
        /// Recency cap; most-recent `limit` entries, newest-first. Defaults to 50.
        #[serde(default = "memory_default_limit")]
        limit: usize,
    },
}

fn memory_empty_metadata() -> Value {
    Value::Object(Map::new())
}

fn memory_default_limit() -> usize {
    50
}

File: typescript/packages/tools/src/params.ts
/**
 * Per-tool input schemas (zod) and a parse helper that maps validation
 * errors to {@link ToolExecutionError} `invalid_parameters`.
 */

import { z } from "zod";
import type { ToolCall } from "@spore/core";

import type { ToolExecutionError } from "./errors.js";

export function parseParams<T>(
  schema: z.ZodType<T>,
  call: ToolCall,
): { ok: true; value: T } | { ok: false; error: ToolExecutionError } {
  const r = schema.safeParse(call.input);
  if (r.success) return { ok: true, value: r.data };
  return {
    ok: false,
    error: { kind: "invalid_parameters", reason: r.error.message },
  };
}

// ---------- Filesystem ----------

export const ReadFileParamsSchema = z.object({ path: z.string() });
export type ReadFileParams = z.infer<typeof ReadFileParamsSchema>;

export const WriteFileParamsSchema = z.object({
  path: z.string(),
  content: z.string(),
  append: z.boolean().default(false),
});
export type WriteFileParams = z.infer<typeof WriteFileParamsSchema>;

export const ListDirParamsSchema = z.object({
  path: z.string(),
  recursive: z.boolean().default(false),
});
export type ListDirParams = z.infer<typeof ListDirParamsSchema>;

export const DeleteFileParamsSchema = z.object({ path: z.string() });
export type DeleteFileParams = z.infer<typeof DeleteFileParamsSchema>;

export const MoveFileParamsSchema = z.object({
  src: z.string(),
  dst: z.string(),
});
export type MoveFileParams = z.infer<typeof MoveFileParamsSchema>;

// ---------- Exec ----------

/**
 * Parameters for the shell-free {@link import("./exec.js").ExecTool}: a program
 * name plus a verbatim argument vector. No shell is involved.
 */
export const ExecParamsSchema = z.object({
  command: z.string(),
  args: z.array(z.string()).default([]),
  /** Timeout in whole seconds. */
  timeout: z.number().int().nonnegative().nullable().optional(),
});
export type ExecParams = z.infer<typeof ExecParamsSchema>;

/**
 * Parameters for the real {@link import("./exec.js").BashCommandTool}: a single
 * shell `script` run via `/bin/sh -c`, with an optional working directory.
 */
export const ShellCommandParamsSchema = z.object({
  script: z.string(),
  working_dir: z.string().nullable().optional(),
  /** Timeout in whole seconds. */
  timeout: z.number().int().nonnegative().nullable().optional(),
});
export type ShellCommandParams = z.infer<typeof ShellCommandParamsSchema>;

export const RunTestsParamsSchema = z.object({
  command: z.string(),
  working_dir: z.string(),
  timeout: z.number().int().nonnegative().nullable().optional(),
});
export type RunTestsParams = z.infer<typeof RunTestsParamsSchema>;

// ---------- Search ----------

export const GrepFilesParamsSchema = z.object({
  pattern: z.string(),
  path: z.string(),
  recursive: z.boolean().default(false),
});
export type GrepFilesParams = z.infer<typeof GrepFilesParamsSchema>;

export const FindFilesParamsSchema = z.object({
  glob: z.string(),
  path: z.string(),
});
export type FindFilesParams = z.infer<typeof FindFilesParamsSchema>;

/** Output mode for the #81 `grep` tool. Defaults to `content`. */
export const GrepOutputModeSchema = z
  .enum(["content", "files_with_matches", "count"])
  .default("content");
export type GrepOutputMode = z.infer<typeof GrepOutputModeSchema>;

/** Parameters for the #81 `grep` tool (regex search with output modes). */
export const GrepParamsSchema = z.object({
  pattern: z.string(),
  path: z.string(),
  recursive: z.boolean().default(false),
  output_mode: GrepOutputModeSchema,
});
export type GrepParams = z.infer<typeof GrepParamsSchema>;

// ---------- EditFile (#81) ----------

export const EditFileParamsSchema = z.object({
  path: z.string(),
  old_string: z.string(),
  new_string: z.string(),
});
export type EditFileParams = z.infer<typeof EditFileParamsSchema>;

// ---------- SendMessage (#81) ----------

export const SendMessageParamsSchema = z.object({ content: z.string() });
export type SendMessageParams = z.infer<typeof SendMessageParamsSchema>;

// ---------- Web (#81) ----------

export const WebFetchParamsSchema = z.object({ url: z.string() });
export type WebFetchParams = z.infer<typeof WebFetchParamsSchema>;

export const WebSearchParamsSchema = z.object({ query: z.string() });
export type WebSearchParams = z.infer<typeof WebSearchParamsSchema>;

// ---------- TodoWrite (#81) ----------

export const TodoStatusSchema = z.enum([
  "completed",
  "in_progress",
  "pending",
]);
export type TodoStatus = z.infer<typeof TodoStatusSchema>;

export const TodoItemSchema = z.object({
  content: z.string(),
  status: TodoStatusSchema,
});
export type TodoItem = z.infer<typeof TodoItemSchema>;

export const TodoWriteParamsSchema = z.object({
  todos: z.array(TodoItemSchema),
});
export type TodoWriteParams = z.infer<typeof TodoWriteParamsSchema>;

// ---------- Tier 3 control tools (#81) ----------

/** `enter_plan_mode` — `context` is optional; defaults to "". */
export const EnterPlanModeParamsSchema = z.object({
  context: z.string().default(""),
});
export type EnterPlanModeParams = z.infer<typeof EnterPlanModeParamsSchema>;

/**
 * `exit_plan_mode` — the agent supplies the structured plan, which deserializes
 * directly into the existing {@link "@spore/core".PlanArtifact} shape (issue
 * #81, Q4a — no stub). `tasks` required; `rationale` defaults to "".
 */
export const ExitPlanModeParamsSchema = z.object({
  plan: z.object({
    tasks: z.array(z.string()),
    rationale: z.string().default(""),
  }),
});
export type ExitPlanModeParams = z.infer<typeof ExitPlanModeParamsSchema>;

/** `ask_user_question` — `question` required; `options` optional. */
export const AskUserQuestionParamsSchema = z.object({
  question: z.string(),
  options: z.array(z.string()).optional(),
});
export type AskUserQuestionParams = z.infer<
  typeof AskUserQuestionParamsSchema
>;

/** `abort` — graceful stop with a `reason`. */
export const AbortParamsSchema = z.object({ reason: z.string() });
export type AbortParams = z.infer<typeof AbortParamsSchema>;

// ---------- Git ----------

export const GitLogParamsSchema = z.object({
  n: z.number().int().nonnegative().default(20),
  format: z.string().default("oneline"),
});
export type GitLogParams = z.infer<typeof GitLogParamsSchema>;

export const GitDiffParamsSchema = z.object({
  from: z.string().nullable().optional(),
  to: z.string().nullable().optional(),
});
export type GitDiffParams = z.infer<typeof GitDiffParamsSchema>;

export const GitCommitParamsSchema = z.object({
  message: z.string(),
  files: z.array(z.string()).default([]),
});
export type GitCommitParams = z.infer<typeof GitCommitParamsSchema>;

export const GitStatusParamsSchema = z.object({}).passthrough();
export type GitStatusParams = z.infer<typeof GitStatusParamsSchema>;

export const GitResetModeSchema = z.enum(["hard", "soft", "mixed"]);
export type GitResetMode = z.infer<typeof GitResetModeSchema>;

export const GitResetParamsSchema = z.object({
  target: z.string(),
  mode: GitResetModeSchema,
});
export type GitResetParams = z.infer<typeof GitResetParamsSchema>;

// ---------- HTTP ----------

export const HttpGetParamsSchema = z.object({
  url: z.string(),
  headers: z.record(z.string()).nullable().optional(),
});
export type HttpGetParams = z.infer<typeof HttpGetParamsSchema>;

export const HttpPostParamsSchema = z.object({
  url: z.string(),
  body: z.unknown(),
  headers: z.record(z.string()).nullable().optional(),
});
export type HttpPostParams = z.infer<typeof HttpPostParamsSchema>;

// ---------- Subagent ----------

export const SubagentParamsSchema = z.object({ instruction: z.string() });
export type SubagentParams = z.infer<typeof SubagentParamsSchema>;

// ---------- TaskList (#71) ----------

const TaskStatusParamSchema = z.enum([
  "pending",
  "in_progress",
  "completed",
  "blocked",
]);

/**
 * Parameters for the {@link import("./tasklist.js").TaskListTool}, a
 * discriminated union on `action`. Each variant carries exactly the fields that
 * action consumes; mirrors the Rust `TaskListParams` enum.
 */
export const TaskListParamsSchema = z.discriminatedUnion("action", [
  z.object({
    action: z.literal("add_task"),
    description: z.string(),
  }),
  z.object({
    action: z.literal("update_task"),
    id: z.number().int().nonnegative(),
    status: TaskStatusParamSchema.optional(),
    description: z.string().optional(),
  }),
  z.object({
    action: z.literal("complete_task"),
    id: z.number().int().nonnegative(),
  }),
  z.object({
    action: z.literal("list_tasks"),
  }),
]);
export type TaskListParams = z.infer<typeof TaskListParamsSchema>;
